File: hello-grpc-python/server/protoServer.py
# encoding: utf-8
import logging
import time
import uuid

import grpc
from concurrent import futures

from landing_pb2 import landing_pb2
from landing_pb2 import landing_pb2_grpc

logger = logging.getLogger('grpc-server')
logger.setLevel(logging.INFO)
console = logging.StreamHandler()
console.setLevel(logging.INFO)
formatter = logging.Formatter('%(asctime)s [%(levelname)s] - %(message)s')
console.setFormatter(formatter)
logger.addHandler(console)

hellos = ["Hello", "Bonjour", "Hola", "こんにちは", "Ciao", "안녕하세요"]


def build_result(data):
    result = landing_pb2.TalkResult()
    result.id = int((time.time()))
    result.type = landing_pb2.OK
    result.kv["id"] = str(uuid.uuid1())
    result.kv["idx"] = data
    index = int(data)
    result.kv["data"] = hellos[index]
    result.kv["meta"] = "PYTHON"
    return result


def read_headers(method_name, context):
    metadata = context.invocation_metadata()
    metadata_dict = {}
    for c in metadata:
        logger.info("%s HEADER %s:%s", method_name, c.key, c.value)
        metadata_dict[c.key] = c.value
    return metadata_dict


class LandingServiceServer(landing_pb2_grpc.LandingServiceServicer):
    def talk(self, request, context):
        read_headers("TALK", context)
        logger.info("TALK REQUEST: data=%s,meta=%s", request.data, request.meta)
        response = landing_pb2.TalkResponse()
        response.status = 200
        result = build_result(request.data)
        response.results.append(result)
        return response

    def talkOneAnswerMore(self, request, context):
        read_headers("TalkOneAnswerMore", context)
        logger.info("TalkOneAnswerMore REQUEST: data=%s,meta=%s", request.data, request.meta)
        datas = request.data.split(",")
        for data in datas:
            response = landing_pb2.TalkResponse()
            response.status = 200
            result = build_result(data)
            response.results.append(result)
            yield response

    def talkMoreAnswerOne(self, request_iterator, context):
        read_headers("TalkMoreAnswerOne", context)
        response = landing_pb2.TalkResponse()
        response.status = 200
        for request in request_iterator:
            logger.info("TalkMoreAnswerOne REQUEST: data=%s,meta=%s", request.data, request.meta)
            response.results.append(build_result(request.data))
        return response

    def talkBidirectional(self, request_iterator, context):
        read_headers("TalkBidirectional", context)
        for request in request_iterator:
            logger.info("TalkBidirectional REQUEST: data=%s,meta=%s", request.data, request.meta)
            response = landing_pb2.TalkResponse()
            response.status = 200
            result = build_result(request.data)
            response.results.append(result)
            yield response


def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    landing_pb2_grpc.add_LandingServiceServicer_to_server(LandingServiceServer(), server)
    server.add_insecure_port('[::]:9996')
    server.start()
    try:
        while True:
            time.sleep(60 * 60 * 24)
    except KeyboardInterrupt:
        server.stop(0)


if __name__ == '__main__':
    serve()

File: hello-grpc-nodejs/proto_server.js
let grpc = require('grpc')
let uuid = require('uuid')
let messages = require('./common/landing_pb')
let services = require('./common/landing_grpc_pb')

let hellos = ["Hello", "Bonjour", "Hola", "こんにちは", "Ciao", "안녕하세요"]

const {createLogger, format, transports} = require('winston')
const {combine, timestamp, printf} = format
const formatter = printf(({level, message, timestamp}) => {
    return `${timestamp} ${level}: ${message}`
})
const logger = createLogger({
    level: 'info',
    format: combine(
        timestamp(),
        formatter
    ),
    transports: [new transports.Console()],
})

function talk(call, callback) {
    printHeaders("Talk", call)
    let response = new messages.TalkResponse()
    response.setStatus(200)
    const talkResult = buildResult(call.request.getData())
    let talkResults = [talkResult]
    response.setResultsList(talkResults)
    logger.info("Talk:" + response)
    callback(null, response)
}

function talkOneAnswerMore(call) {
    printHeaders("TalkOneAnswerMore",call)
    let datas = call.request.getData().split(",")
    for (const data in datas) {
        let response = new messages.TalkResponse()
        response.setStatus(200)
        const talkResult = buildResult(data)
        let talkResults = [talkResult]
        response.setResultsList(talkResults)
        logger.info("TalkOneAnswerMore:" + response)
        call.write(response)
    }
    call.end()
}

function talkMoreAnswerOne(call, callback) {
    printHeaders("TalkMoreAnswerOne",call)
    let talkResults = []
    call.on('data', function (request) {
        talkResults.push(buildResult(request.getData()))
    })
    call.on('end', function () {
        let response = new messages.TalkResponse()
        response.setStatus(200)
        response.setResultsList(talkResults)
        logger.info("TalkMoreAnswerOne:" + response)
        callback(null, response)
    })
}

function talkBidirectional(call) {
    printHeaders("TalkBidirectional",call)
    call.on('data', function (request) {
        let response = new messages.TalkResponse()
        response.setStatus(200)
        let data = request.getData();
        const talkResult = buildResult(data)
        let talkResults = [talkResult]
        response.setResultsList(talkResults)
        logger.info("TalkBidirectional:" + response)
        call.write(response)
    })
    call.on('end', function () {
        call.end()
    })
}

// {"status":200,"results":[{"id":1600402320493411000,"kv":{"data":"Hello","id":"0"}}]}
function buildResult(id) {
    let result = new messages.TalkResult()
    let index = parseInt(id)
    result.setId(Math.round(Date.now() / 1000))
    result.setType(messages.ResultType.OK)
    let kv = result.getKvMap()
    kv.set("id", uuid.v1())
    kv.set("idx", id)
    kv.set("data", hellos[index])
    kv.set("meta", "NODEJS")
    return result
}

/**
 * Starts an RPC server that receives requests for the Greeter service at the
 * sample server port
 */
function main() {
    let server = new grpc.Server()
    server.addService(services.LandingServiceService, {
        talk: talk,
        talkOneAnswerMore: talkOneAnswerMore,
        talkMoreAnswerOne: talkMoreAnswerOne,
        talkBidirectional: talkBidirectional
    })
    server.bind('0.0.0.0:9996', grpc.ServerCredentials.createInsecure())
    server.start()
}

function printHeaders(methodName,call) {
    let headers = call.metadata.getMap();
    for (let key in headers) {
        logger.info(methodName+ " HEADER: " + key + ":" + headers[key])
    }
}

main()